File: python/mujoco_mpc/kalman_test.py
# Copyright 2023 DeepMind Technologies Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

from absl.testing import absltest
import mujoco
from mujoco_mpc import kalman as kalman_lib
import numpy as np

import pathlib


class KALMANTest(absltest.TestCase):

  def test_settings(self):
    # load model
    model_path = (
        pathlib.Path(__file__).parent.parent.parent
        / "mjpc/test/testdata/estimator/particle/task1D.xml"
    )
    model = mujoco.MjModel.from_xml_path(str(model_path))

    # initialize
    kalman = kalman_lib.KALMAN(model=model)

    # settings
    epsilon = 2.0
    flg_centered = True
    auto_timestep = True
    settings = kalman.settings(
        epsilon=epsilon, flg_centered=flg_centered, auto_timestep=auto_timestep
    )

    # test
    self.assertLess(np.abs(settings["epsilon"] - epsilon), 1.0e-5)
    self.assertTrue(settings["flg_centered"] == flg_centered)
    self.assertTrue(settings["auto_timestep"] == auto_timestep)

  def test_updates(self):
    # load model
    model_path = (
        pathlib.Path(__file__).parent.parent.parent
        / "mjpc/test/testdata/estimator/particle/task1D.xml"
    )
    model = mujoco.MjModel.from_xml_path(str(model_path))

    # initialize
    kalman = kalman_lib.KALMAN(model=model)

    # state
    state = np.random.normal(scale=1.0, size=(model.nq + model.nv))
    state_response = kalman.state(state=state)

    # test state
    self.assertLess(np.linalg.norm(state_response - state), 1.0e-5)

    # covariance
    nvelocity = 2 * model.nv
    F = np.random.normal(scale=1.0, size=(nvelocity**2)).reshape(nvelocity, nvelocity)
    covariance = F.T @ F
    covariance_response = kalman.covariance(covariance=covariance)

    # test covariance
    self.assertLess(np.linalg.norm((covariance_response - covariance).ravel()), 1.0e-5)
    self.assertTrue(covariance_response.shape == (nvelocity, nvelocity))

    # noise
    process = np.random.normal(scale=1.0e-3, size=nvelocity)
    sensor = np.random.normal(scale=1.0e-3, size=model.nsensordata)
    noise = kalman.noise(process=process, sensor=sensor)

    # test noise
    self.assertLess(np.linalg.norm(noise["process"] - process), 1.0e-5)
    self.assertLess(np.linalg.norm(noise["sensor"] - sensor), 1.0e-5)

    # measurement update
    ctrl = np.random.normal(scale=1.0, size=model.nu)
    sensor = np.random.normal(scale=1.0, size=model.nsensordata)
    kalman.update_measurement(ctrl=ctrl, sensor=sensor)

    # # prediction update
    kalman.update_prediction()


    # timers
    timer = kalman.timers()

    self.assertTrue(timer["measurement"] > 0.0)
    self.assertTrue(timer["prediction"] > 0.0)

if __name__ == "__main__":
  absltest.main()

File: mjpc/test/estimator/kalman_test.cc
// Copyright 2023 DeepMind Technologies Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "mjpc/estimators/kalman.h"

#include <vector>

#include <absl/random/random.h>
#include <mujoco/mujoco.h>

#include "gtest/gtest.h"
#include "mjpc/direct/trajectory.h"
#include "mjpc/test/load.h"
#include "mjpc/test/simulation.h"
#include "mjpc/utilities.h"

namespace mjpc {
namespace {

TEST(Estimator, Kalman) {
  // load model
  mjModel* model = LoadTestModel("estimator/particle/task_imu.xml");
  mjData* data = mj_makeData(model);

  // ----- rollout ----- //
  int T = 200;
  Simulation sim(model, T);
  auto controller = [](double* ctrl, double time) {};
  double qpos0[1] = {0.25};
  sim.SetState(qpos0, NULL);
  sim.Rollout(controller);

  // ----- Kalman ----- //

  // initialize filter
  Kalman kalman(model);

  // set initial state
  mju_copy(kalman.state.data(), sim.qpos.Get(0), model->nq);
  mju_copy(kalman.state.data() + model->nq, sim.qvel.Get(0), model->nv);

  // initialize covariance
  mju_eye(kalman.covariance.data(), 2 * model->nv);
  mju_scl(kalman.covariance.data(), kalman.covariance.data(), 1.0e-5,
          (2 * model->nv) * (2 * model->nv));

  // initial process noise
  mju_fill(kalman.noise_process.data(), 1.0e-5, 2 * model->nv);

  // initialize sensor noise
  mju_fill(kalman.noise_sensor.data(), 1.0e-5, model->nsensordata);

  // Kalman trajectories
  DirectTrajectory<double> kalman_qpos(model->nq, T);
  DirectTrajectory<double> kalman_qvel(model->nv, T);
  DirectTrajectory<double> kalman_timer_measurement(1, T);
  DirectTrajectory<double> kalman_timer_prediction(1, T);

  // noisy sensor
  std::vector<double> noisy_sensor(model->nsensordata);
  absl::BitGen gen_;

  for (int t = 0; t < T; t++) {
    // noisy sensor
    mju_copy(noisy_sensor.data(), sim.sensor.Get(t), model->nsensordata);
    for (int i = 0; i < model->nsensordata; i++) {
      noisy_sensor[i] += 0.0 * absl::Gaussian<double>(gen_, 0.0, 1.0);
    }
    // measurement update
    kalman.UpdateMeasurement(sim.ctrl.Get(t), noisy_sensor.data());

    // cache state
    kalman_qpos.Set(kalman.state.data(), t);
    kalman_qvel.Set(kalman.state.data() + model->nq, t);

    // cache timer
    double timer_measurement = kalman.TimerMeasurement();
    kalman_timer_measurement.Set(&timer_measurement, t);

    // prediction update
    kalman.UpdatePrediction();

    // cache timer
    double timer_prediction = kalman.TimerPrediction();
    kalman_timer_prediction.Set(&timer_prediction, t);

    // test qpos
    std::vector<double> pos_error(model->nq);
    mju_sub(pos_error.data(), kalman_qpos.Get(t), sim.qpos.Get(t), model->nq);
    EXPECT_NEAR(mju_norm(pos_error.data(), model->nq), 0.0, 1.0e-4);

    // test qvel
    std::vector<double> vel_error(model->nv);
    mju_sub(vel_error.data(), kalman_qvel.Get(t), sim.qvel.Get(t), model->nv);
    EXPECT_NEAR(mju_norm(vel_error.data(), model->nv), 0.0, 1.0e-4);
  }

  // delete data + model
  mj_deleteData(data);
  mj_deleteModel(model);
}

}  // namespace
}  // namespace mjpc
